File: python/trees/rightSideView.py
from collections import deque


class Solution:
    def rightSideViewBFS(self, root):
        if not root:
            return []
        result = []
        queue = deque([root])

        while queue:
            queueLen = len(queue)
            for i in range(queueLen):
                current = queue.popleft()
                if i == queueLen - 1:
                    result.append(current.val)

                if (current.left):
                    queue.append(current.left)
                if current.right:
                    queue.append(current.right)

        return result

    def rightSideViewDFS(self, root):

File: js/src/trees/rightSideView.js
const rightSideView = (root) => {
  const response = [];
  const queue = [root];

  while (queue.length) {
    const queueLen = queue.length;
    for (let i = 0; i < queueLen.length; i++) {
      const current = queue.shift();
      if (queueLen.length - 1) {
        response.push(current.val);
      }
      if (current.left) {
        queue.push(current.left);
      }
      if (current.right) {
        queue.push(current.right);
      }
    }
  }
  return response;
};

const rightSideViewDfs = (root) => {
  const response = new Map();

  const dfs = (root, level = 0) => {
    if (!root) return;

    if (!response.has(level)) response.set(level, root.val);

    if (root.right) {
      dfs(root.left, level + 1);
    }
    if (root.left) {
      dfs(root.right, level + 1);
    }
  };

  dfs(root);
  return response;
};
